### ISSUE TITLE
i18n on custom elements with @bindable does not work with html attributes

### ISSUE THREAD

**Issue by DerAlbertCom** [aurelia/i18n#123]
attributes on an HTML Tag can be localized with the attribute  `<div  i18n="[attribute]key" />` works on plain html tags, but not on custom elements.

On the view model i have `@bindable title` and while using `<custom i18n="[title]thekey"></custom>` title gets not translated.  i know i can to `<custom title="${'theKey' | t}"><custom>` but this inconsistent.

Is this a bug, a feature, planned?

Any hint for self fixing this? Which can result in a PR. Or are reasons behind not allowing this?

**Comment by EisenbergEffect** [comment]
@zewa666 This sounds like a bug in the way that the i18n library is setting attributes when they relate to a custom attribute. In this case, they need to be set on the view-model instance, not on the html element itself. @jdanyow can likely advise you on the best way to handle this. He had to implement some similar fixes recently when working on validation.

**Comment by jdanyow** [comment]
here's the code aurelia-validation uses to retrieve the DOM element associated with a binding:

https://github.com/aurelia/validation/blob/0.11.0/src/validate-binding-behavior.js#L12-L31

**Comment by jdanyow** [comment]
@DerAlbertCom - I want to make sure I understand the question... is this what you are looking for:

```js
export class MyCustomElement {
  @bindable title;
}
```

```html
<template>
  <my-custom-element title.bind="myProperty" i18n="[title]thekey"></my-custom-element>

  <my-custom-element title="${myProperty}" i18n="[title]thekey"></my-custom-element>

  <my-custom-element title="hello I'm a title" i18n="[title]thekey"></my-custom-element>
</template>
```

In other words, you want the i18n attribute to localize the bound value of the custom element's title property, *without using i18n in the title binding directly*.

Is that correct?

**Comment by DerAlbertCom** [comment]
Now i don't know if understand you @jdanyow 

But i think yes ;)

i would like to localize a bindable property of a custom element with the *i18n* attribute from the i18n plugin. In this sample the title property/attribute. Like we can do with plain html elements.

```HTML
<template>
  <my-custom-element i18n="[title]thekey" i18n-params.bind="whatever"></my-custom-element>
  <my-custom-element i18n="[title]thekey"></my-custom-element>
</template>
```

**Comment by zewa666** [comment]
I'm still not sure whether this is a good fit tbh. What you're doing this way is manipulating the component from outside like e.g. passing on styles to the component. Later is nice and fine and in worst case the rendering could look strange, but in case of translations it could interfere with the components internals.

So typically I would see the requested component using the i18n attribute by itself where it is appropriate. You could then create custom attributes, which instead of the values accept keys and you forward them further on to the translation.

Long story short, I think the translation feature of a component is more of a business/component logic  vs styling, as such I personally feel like it should be handled by the component itself vs attached atop from the caller. Nevertheless I'm open to discussion if you have specific use cases.

**Comment by DerAlbertCom** [comment]
Because it is business logic it should not be done within the component.

My use case is that we have a `Toolbar`. This toolbar is used in different views and it display some information. These Information is in structurel the same, but the semantic and context is different.

e.g. On one view we display "{{selectedCount}} books selected", on  a other "{{selectedCount} users selected" (and more text is used real, only as an example here).

The Component has the functionality, but does not know the context. So we have to pass the context in to the component. Thats why i18n should work the same on components as on tags. It passing context to the tag or component.

**Comment by zewa666** [comment]
The nice thing with what belongs where is that nobody is right and nobody is wrong :)

I can see your point though, although I personally would prefer something else. So lets put this as an enhancement. In the end it can't harm and nobody is forced to use only one approach.

**Comment by zewa666** [comment]
@DerAlbertCom the recent release contains this feature. Please give it a try and reply whether its what you thought of

**Comment by zewa666** [comment]
@tipa thanks for the hint, a patch is checked in and should be released soon

**Comment by zewa666** [comment]
related issue: https://github.com/aurelia/i18n/issues/146

**Comment by zewa666** [comment]
fixed @tipa and @DerAlbertCom, ready for 2nd try :)

**Comment by DerAlbertCom** [comment]
@zewa666 currently on vacation :tropical_drink:  but thank you for implementing this.

**Comment by zewa666** [comment]
@DerAlbertCom I'm gonna close this one for now, if you experience any issues we can reopen it

**Comment by DerAlbertCom** [comment]
is this merged in 1.1.2?

**Comment by zewa666** [comment]
Yep it is

**Comment by gronostajo** [comment]
I'm affected by this too. I'm using Aurelia-I18N 1.2.3 and this doesn't work as expected:

```
<multilingual-text t="[label]resourcesConfig.metadata.column.label"></multilingual-text>
```

Aurelia Context shows following properties:

```
label: undefined
value: "[label]resourcesConfig.metadata.column.label"
```

**Comment by zewa666** [comment]
@gronostajo can I ask you for a Link to an example so I can take a look at the issue?

**Comment by gronostajo** [comment]
Sure. Here's a [minimal diff to reproduce the problem](https://github.com/gronostajo/aurelia-i18n-bug/commit/5979156426cacd634bc8631c4a71cc6185c98530) and a [demo](https://gronostajo.github.io/aurelia-i18n-bug/). For me it looks like this:

![image](https://cloud.githubusercontent.com/assets/1347866/21206177/a51ab098-c260-11e6-9962-66f78ed5836b.png)

**Comment by zewa666** [comment]
oh man what an embarrassing issue, I've been [checking the value](https://github.com/aurelia/i18n/commit/2c45b08c8e8964131bcab4259c8474ff26042b42#diff-1e8d6c22d70a5ac18d47adf06afa1efeR241) instead of the presence of the attribute/bindable. Sorry about that.

Thanks for the nice example, this really helped me to get down quickly to the issue.
The fix will be out with the next release, in the meantime if your project depends on it, just initialize the bindable with something ... empty string for example :)
